From 36cd6957060db87e610ef15d8199fb6d8e025110 Mon Sep 17 00:00:00 2001
From: Oliver Pinter <oliver.pntr@gmail.com>
Date: Wed, 1 May 2013 10:12:22 +0200
Subject: [PATCH 2/2] PaX ASLR - separate shared page from stack

in special case the shared page ontop of user memory become RWX, after
this patch, this behavior elliminated

before the patch:
op@opn ~> procstat -v `pgrep firefox` | tail -2
60839 0x7ffffffdf000 0x7ffffffff000 rwx   32    0   1   0 C--- df
60839 0x7ffffffff000 0x800000000000 rwx    1    0  79   0 CN-- ph

after the patch:
2644 0x7ffffffdb000 0x7fffffffb000 rwx   22    0   1   0 C--D df
2644 0x7ffffffff000 0x800000000000 r-x    1    0  55   0 CN-- ph

Signed-off-by: Oliver Pinter <oliver.pntr@gmail.com>

PaX ASLR - separate shared page from stack in linuxulator

Signed-off-by: Oliver Pinter <oliver.pntr@gmail.com>
---
 sys/amd64/include/vmparam.h | 2 +-
 sys/amd64/linux32/linux.h   | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/sys/amd64/include/vmparam.h b/sys/amd64/include/vmparam.h
index e06fa39..1bc57c3 100644
--- a/sys/amd64/include/vmparam.h
+++ b/sys/amd64/include/vmparam.h
@@ -176,7 +176,7 @@
 #define	VM_MAXUSER_ADDRESS	UVADDR(NUPML4E, 0, 0, 0)
 
 #define	SHAREDPAGE		(VM_MAXUSER_ADDRESS - PAGE_SIZE)
-#define	USRSTACK		SHAREDPAGE
+#define	USRSTACK		(SHAREDPAGE - 4*PAGE_SIZE)
 
 #define	VM_MAX_ADDRESS		UPT_MAX_ADDRESS
 #define	VM_MIN_ADDRESS		(0)
diff --git a/sys/amd64/linux32/linux.h b/sys/amd64/linux32/linux.h
index 7b52a64..1ba0dea 100644
--- a/sys/amd64/linux32/linux.h
+++ b/sys/amd64/linux32/linux.h
@@ -50,7 +50,7 @@ MALLOC_DECLARE(M_LINUX);
 
 #define	LINUX32_MAXUSER		((1ul << 32) - PAGE_SIZE)
 #define	LINUX32_SHAREDPAGE	(LINUX32_MAXUSER - PAGE_SIZE)
-#define	LINUX32_USRSTACK	LINUX32_SHAREDPAGE
+#define	LINUX32_USRSTACK	(LINUX32_SHAREDPAGE - 4*PAGE_SIZE)
 
 /* XXX 16 = sizeof(linux32_ps_strings) */
 #define	LINUX32_PS_STRINGS	(LINUX32_USRSTACK - 16)
-- 
1.8.2

